swift/ql/test/query-tests/Security/CWE-079/UnsafeWebViewFetch.swift

```diff
@@ -0,0 +1,217 @@
+
+class NSObject
+{
+}
+
+class URL
+{
+	init?(string: String) {}
+	init?(string: String, relativeTo: URL?) {}
+}
+
+extension String {
+	init(contentsOf: URL) throws {
+        var data = ""
+        
+        // ...
+        
+        self.init(data)
+    }
+}
+
+class NSURLRequest : NSObject
+{
+	enum CachePolicy : UInt
+	{
+		case useProtocolCachePolicy
+	}
+}
+
+typealias TimeInterval = Double
+
+class URLRequest
+{
+	typealias CachePolicy = NSURLRequest.CachePolicy
+
+	init(url: URL, cachePolicy: URLRequest.CachePolicy = .useProtocolCachePolicy, timeoutInterval: TimeInterval = 60.0) {}
+}
+
+class Data
+{
+    init<S>(_ elements: S) {}
+}
+
+class WKNavigation : NSObject
+{
+}
+
+class UIResponder : NSObject
+{
+}
+
+class UIView : UIResponder
+{
+}
+
+class UIWebView : UIView
+{
+	func loadRequest(_ request: URLRequest) {} // deprecated
+
+	func load(_ data: Data, mimeType MIMEType: String, textEncodingName: String, baseURL: URL) {} // deprecated
+
+	func loadHTMLString(_ string: String, baseURL: URL?) {} // deprecated
+}
+
+class WKWebView : UIView
+{
+	func load(_ request: URLRequest) -> WKNavigation? {
+		// ...
+
+		return WKNavigation()
+	}
+
+	func load(_ data: Data, mimeType MIMEType: String, characterEncodingName: String, baseURL: URL) -> WKNavigation? {
+		// ...
+
+		return WKNavigation()
+	}
+
+	func loadHTMLString(_ string: String, baseURL: URL?) -> WKNavigation? {
+		// ...
+
+		return WKNavigation()
+	}
+}
+
+// --- tests ---
+
+func getRemoteData() -> String {
+	let url = URL(string: "http://example.com/")
+	do
+	{
+		return try String(contentsOf: url!)
+	} catch {
+		return ""
+	}
+}
+
+func testSimpleFlows() {
+	let webview = UIWebView()
+
+	webview.loadHTMLString(try! String(contentsOf: URL(string: "http://example.com/")!), baseURL: nil) // BAD [NOT DETECTED]
+
+	let data = try! String(contentsOf: URL(string: "http://example.com/")!)
+	webview.loadHTMLString(data, baseURL: nil) // BAD [NOT DETECTED]
+
+	let url = URL(string: "http://example.com/")
+	webview.loadHTMLString(try! String(contentsOf: url!), baseURL: nil) // BAD [NOT DETECTED]
+}
+
+func testUIWebView() {
+	let webview = UIWebView()
+
+	let localString = "<html><body><p>Local HTML</p></body></html>"
+	let localStringFragment = "<body><p>Local HTML</p></body>"
+	let remoteString = getRemoteData()
+
+	webview.loadHTMLString(localString, baseURL: nil) // GOOD: the HTML data is local
+	webview.loadHTMLString(getRemoteData(), baseURL: nil) // BAD: HTML contains remote input, may access local secrets
+	webview.loadHTMLString(remoteString, baseURL: nil) // BAD
+
+	webview.loadHTMLString("<html>" + localStringFragment + "</html>", baseURL: nil) // GOOD: the HTML data is local
+	webview.loadHTMLString("<html>" + remoteString + "</html>", baseURL: nil) // BAD
+
+	webview.loadHTMLString("<html>\(localStringFragment)</html>", baseURL: nil) // GOOD: the HTML data is local
+	webview.loadHTMLString("<html>\(remoteString)</html>", baseURL: nil) // BAD [NOT DETECTED]
```

Comment:
I'm not surprised this isn't caught, currently. We do extract the relevant parts, but it's not hooked up properly in the CFG/dataflow library yet ([here's](https://github.com/github/codeql-c-team/issues/1078) the internal issue for this).

Once that library issue is resolved, I imagine this missing flow will appear in this query.

Reply:
I've made it part of https://github.com/github/codeql-c-team/issues/1108, though obviously it depends on the issue you cited.